network/src/consensus.rs

```diff
@@ -167,9 +167,9 @@ impl<P, E> Network for ConsensusNetwork<P,E> where
 					knowledge,
 					local_session_key,
 				});
-				let inner_stream = match rx.poll() {
-					Ok(Async::Ready(inner_stream)) => inner_stream,
-					_ => unreachable!("1. The with_gossip closure executed first, 2. the reply to the oneshot should be available")
+				let inner_stream = match rx.try_recv() {
+					Ok(inner_stream) => inner_stream,
+					_ => unreachable!("1. The with_gossip closure executed first, 2. the reply should be available")
```

Comment:
i guess I don't understand why the `with_gossip` function doesn't just wrap this logic if the guarantees are strong enough that it will always return something.

Reply:
(also doesn't `try_recv` race against `with_gossip` actually executing the provided closure?

Reply:
> i guess I don't understand why the `with_gossip` function doesn't just wrap this logic if the guarantees are strong enough that it will always return something.

If I understand correctly what you mean, the second `with_spec` call cannot be made inside the `with_gossip` closure, since that would require moving `self.network` into it, and moving it in another thread(I tried, it seems self.network is not thread-safe).

> also doesn't `try_recv` race against `with_gossip` actually executing the provided closure

Both the `with_gossip` and `with_spec` closure end up being sent in a message, and handled sequentially on the protocol thread. Since that thread only handles one message at a time, we can be sure that the closures will execute one after the other. So by the time the `try_recv` is called, the first closure should have executed already.